src/arduino.cc/builder/ctags_target_file_saver.go

```diff
@@ -59,3 +67,53 @@ func (s *CTagsTargetFileSaver) Run(context map[string]interface{}) error {
 
 	return nil
 }
+
+func saveLinesContainingDirectivesAndSketch(src string, tofind []string) string {
+	lines := strings.Split(src, "\n")
+
+	saveLine := false
+	minimizedString := ""
+
+	for _, line := range lines {
+		if (saveLine || startsWithHashtag(line)) {
+			minimizedString += line + "\n"
+		}
+		if (containsAny(line, tofind)) {
+			saveLine = true
+		}
+		if (saveLine && !containsAny(line, tofind) && isLineDirective(line)) {
+			saveLine = false
+		}
+	}
+	return minimizedString
+}
+
+
+func containsAny(src string, tofind []string) bool {
+	for _, str := range tofind {
+		if strings.Contains(src, str) {
+			return true
+		}
+	}
+	return false
+}
+
+func startsWithHashtag(src string) bool {
+	trimmedStr := strings.TrimSpace(src)
+	if (len(trimmedStr) > 0 && trimmedStr[0] == '#') {
+		return true
+	}
+	return false
+}
+
+func isLineDirective(src string) bool {
```

Comment:
It seems this function is slightly mis-named. I would expect it checks for [#line directives](https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html), but it seems it actually checks for ["line markers" generated by the preprocessor](https://gcc.gnu.org/onlinedocs/cpp/Preprocessor-Output.html). I guess changing the name should be changed correspondingly?

Also, it seems that any non-printing characters in the filenames will be escaped using \xxx escapes by the preprocessor. The current substring-based matching above doesn't handle these properly. I think the correct way to handle this, is to make `isLineDirective` (which should probably be renamed to `filenameFromLineMarker` or something?) actually parse the line marker, apply unescaping to the filename and return that?